content/articles/working-with-java-synchronization/index.md

```diff
@@ -0,0 +1,298 @@
+                @Override
+                public void run() {
+                    synchronized (zy) {
+                        zy.incrementCheck();
+                        System.out.println("The output of the thread is : " + Thread.currentThread().getName() + " - " + zy.getCheck());
+                    }
+                }
+            }).start();
+        }
+    }
+}
+```
+
+```
+The output of the thread is : Thread-0 - 2
+The output of the thread is : Thread-4 - 3
+The output of the thread is : Thread-3 - 4
+The output of the thread is : Thread-2 - 5
+The output of the thread is : Thread-1 - 6
+```
+
+From the results of the threads above, no thread shares a value. Our aim of avoiding a race condition is fulfilled.
+
+### Understanding synchronized methods and synchronized blocks
+To understand the two, let us first look at each separately:
+
+#### Synchronized methods
+These consist of the following properties:
+
+- A synchronized method locks down the whole object. No other thread may access synchronized methods in the object while the method is executing. With static methods, they are locked by their class.
+- Synchronized methods need locks on the current object or, in the case of static methods, the whole class. This is because the lock is obtained when the thread enters and released when the thread quits (naturally or by throwing an exception).
+- Synchronized methods keep the lock for the method scope.
+- A synchronized static function prevents instances from being changed. 
+
+#### Synchronized blocks
+The synchronized block may be used to perform synchronization on any specified resource of the method.
+
+Suppose we have 100 lines of code in our method, but we want to synchronize only 10 lines, in such cases, we can use synchronized block. If we put all the codes of the method in the synchronized block, it will work same as the synchronized method.
+
+Let us look at an example program that uses synchronized blocks.
+
+Save it as `TestSynchronizedBlock1.java`
+
+```java
+class Table
+{
+    void printTable(int n){
+        synchronized(this){//synchronized block
+            for(int i=1;i<=5;i++){
+                System.out.println(n*i);
+                try{
+                    Thread.sleep(10);
+                }catch(Exception e){System.out.println(e);}
+            }
+        }
+    }//end of the method
+}
+
+class MyThread1 extends Thread{
+    Table t;
+    MyThread1(Table t){
+        this.t=t;
+    }
+    public void run(){
+        t.printTable(5);
+    }
+
+}
+class MyThread2 extends Thread{
+    Table t;
+    MyThread2(Table t){
+        this.t=t;
+    }
+    public void run(){
+        t.printTable(100);
+    }
+}
+
+public class TestSynchronizedBlock1{
+    public static void main(String[] args){
+        Table obj = new Table();//only one object
+        MyThread1 t1=new MyThread1(obj);
+        MyThread2 t2=new MyThread2(obj);
+        t1.start();
+        t2.start();
+    }
+}

```

Comment:
There still seems to be a race condition here. Most of the time I get the correct output but occasionally I get this:
![Capture](https://user-images.githubusercontent.com/63324129/163846281-a8b12ff6-4ddc-43f2-a584-9132936fb53f.PNG)